File: backend/Domain/Entities/Policy.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Domain.Enums;

namespace Domain.Entities
{
    public class Policy
    {
        [Key]
        public Guid Id { get; set; } = Guid.NewGuid();

        [Required]
        public Guid UserId { get; set; } // Customer

        [Required]
        public Guid PlanId { get; set; } // Plan

        public Guid? AgentId { get; set; } // Agent assigned

        [ForeignKey(nameof(AgentId))]
        public User? Agent { get; set; }

        public DateTime StartDate { get; set; } = DateTime.UtcNow;
        public DateTime EndDate { get; set; }

        [Required]
        public int DurationInMonths { get; set; }

        [Required]
        public PaymentFrequency PaymentFrequency { get; set; }

        [Required]
        public PolicyStatus Status { get; set; } = PolicyStatus.Active;

        [Required]
        [Range(0, double.MaxValue)]
        public decimal TotalPremium { get; set; }

        [Required]
        [Range(0, double.MaxValue)]
        public decimal TotalPaid { get; set; }

        /// <summary>
        /// Snapshot of the plan's base coverage amount at the time of policy creation.
        /// Does NOT change when the plan is updated.
        /// </summary>
        [Required]
        [Range(0, double.MaxValue)]
        public decimal PlanBaseCoverageAmount { get; set; }

        /// <summary>
        /// Snapshot of the plan's base premium amount at the time of policy creation.
        /// Does NOT change when the plan is updated.
        /// </summary>
        [Required]
        [Range(0, double.MaxValue)]
        public decimal PlanBasePremiumAmount { get; set; }

        /// <summary>
        /// The actual calculated coverage for this specific policy.
        /// Formula: PlanBaseCoverageAmount * (DurationInMonths / Plan.DurationInMonths)
        /// Frozen after creation.
        /// </summary>
        [Required]
        [Range(0, double.MaxValue)]
        public decimal CoverageAmount { get; set; }

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        public User User { get; set; } = null!;
        public Plan Plan { get; set; } = null!;
        public ICollection<PolicyPayment> Payments { get; set; } = new List<PolicyPayment>();
        public ICollection<Claim> Claims { get; set; } = new List<Claim>();
    }
}

File: frontend/src/app/models/policy/policy.ts
import { PlanDto } from './plan';
import { PolicyPaymentDto } from '../payment/payment';

export interface PolicyDto {
    id: string; // Guid
    startDate: string; // DateTime
    endDate: string; // DateTime
    durationInMonths: number;
    paymentFrequency: string;
    status: string;
    totalPremium: number;
    totalPaid: number;
    /** Calculated coverage for this specific policy (frozen at purchase time) */
    coverageAmount: number;
    /** Snapshot of plan's base coverage at time of purchase */
    planBaseCoverageAmount: number;
    /** Snapshot of plan's base premium at time of purchase */
    planBasePremiumAmount: number;
    plan: PlanDto;
    customerName: string;
    agentName?: string;
    payments: PolicyPaymentDto[];
}

export interface PurchasePolicyRequest {
    planId: string;
    durationInYears: number;
    paymentFrequency: number; // 0=Monthly, 1=Quarterly, 2=Yearly
}
